File: packages/domain/src/escrow.ts
/**
 * When the money moves.
 *
 * `pricing.ts` decides what a trip is worth. This decides *when* each part of
 * it changes hands, which is a different question and the one that actually
 * blocks trade between strangers: a carrier will not load without an advance,
 * a shipper will not pay in full before delivery, and neither will go first
 * with somebody they have never met.
 *
 * Milestones make that tractable. Each one has a **condition the platform can
 * verify from evidence it already holds** — not from somebody saying it
 * happened. A milestone that releases on a phone call is a milestone that
 * releases on an argument.
 */

import { percent, subtract, type Kobo } from './money.ts';
import type { TripState } from './trip.ts';

export type MilestoneKind = 'advance' | 'in_transit' | 'delivered' | 'retention';

export interface Milestone {
  readonly kind: MilestoneKind;
  /** Share of the fare. The four sum to 100. */
  readonly pct: number;
  /** What must be true, in words both parties read the same way. */
  readonly condition: string;
}

/**
 * The default schedule.
 *
 * 30 on loading, 20 on the road, 40 on delivery, 10 held.
 *
 * The advance is real money on the day the truck loads, because that is what
 * the diesel is bought with — a schedule that pays nothing until delivery is a
 * schedule only carriers with working capital can accept, which is not the
 * market this exists for.
 *
 * The retention is the smallest number that still means anything. It exists so
 * a shortage discovered at the market has something to be settled against; at
 * 25% it would be a carrier financing the shipper's counting.
 */
export const SCHEDULE: readonly Milestone[] = [
  {
    kind: 'advance',
    pct: 30,
    condition: 'The truck reached the depot and loading started.',
  },
  {
    kind: 'in_transit',
    pct: 20,
    condition: 'The trip has been moving with positions arriving for six hours.',
  },
  {
    kind: 'delivered',
    pct: 40,
    condition: 'Proof of delivery captured: photographs, a signature and a name.',
  },
  {
    kind: 'retention',
    pct: 10,
    condition: 'Seven days after delivery with no exception raised.',
  },
];

/** How long the retention is held. */
export const RETENTION_DAYS = 7;

/** Six hours of arriving positions before the second milestone releases. */
export const IN_TRANSIT_MS = 6 * 60 * 60_000;

/**
 * What the platform knows, at the moment a release is being decided.
 *
 * Named `EscrowConditions` rather than `Conditions` because `tracking.ts`
 * already has the latter — the fourth such collision in this package. ADR-0011
 * says what to do about it.
 */
export interface EscrowConditions {
  readonly state: TripState;
  /** How long the trip has been in transit with positions arriving. */
  readonly movingForMs: number;
  readonly podSealed: boolean;
  readonly deliveredAt: Date | null;
  readonly exceptionRaised: boolean;
}

/**
 * Whether a milestone's condition is met, from evidence the platform holds.
 *
 * Every branch reads something the tracker, the trip machine or the proof
 * engine produced. None of them reads an opinion.
 */
export function isMet(kind: MilestoneKind, conditions: EscrowConditions, now: Date): boolean {
  const started =
    conditions.state !== 'open' && conditions.state !== 'assigned';

  switch (kind) {
    case 'advance':
      return started;

    case 'in_transit':
      return started && conditions.movingForMs >= IN_TRANSIT_MS;

    case 'delivered':
      // Not the `delivered` state — the *proof*. A state is a claim somebody
      // made; the proof is photographs, a signature and a position.
      return conditions.podSealed;

    case 'retention': {
      if (conditions.deliveredAt === null || !conditions.podSealed) return false;
      // An open exception holds the retention. That is the entire reason it
      // exists, and releasing it on a timer regardless would make it theatre.
      if (conditions.exceptionRaised) return false;
      const elapsed = now.getTime() - conditions.deliveredAt.getTime();
      return elapsed >= RETENTION_DAYS * 86_400_000;
    }
  }
}

export interface Release {
  readonly milestone: Milestone;
  readonly amount: Kobo;
  readonly met: boolean;
}

/**
 * The whole schedule against a trip, with what each part is worth.
 *
 * Returns every milestone, met or not. A schedule that showed only what has
 * been released would answer "how much have I had" and never "when do I get
 * the rest", which is the question a carrier is actually asking.
 */
export function schedule(
  agreed: Kobo,
  conditions: EscrowConditions,
  now: Date,
): readonly Release[] {
  return SCHEDULE.map((milestone) => ({
    milestone,
    amount: percent(agreed, milestone.pct),
    met: isMet(milestone.kind, conditions, now),
  }));
}

export function released(releases: readonly Release[]): Kobo {
  return releases
    .filter((release) => release.met)
    .reduce((total, release) => (total + release.amount) as Kobo, 0 as Kobo);
}

export function heldBack(agreed: Kobo, releases: readonly Release[]): Kobo {
  return subtract(agreed, released(releases));
}

/**
 * The next thing that has to happen for money to move.
 *
 * The sentence a carrier wants on the screen. Null when everything has been
 * released — and then the screen says that instead, rather than an empty space
 * where a next step used to be.
 */
export function nextRelease(releases: readonly Release[]): Release | null {
  return releases.find((release) => !release.met) ?? null;
}

/**
 * Whether the schedule adds up.
 *
 * Asserted rather than assumed: a schedule that sums to 95 quietly keeps 5% of
 * every trip, and nobody would notice for months.
 */
export function sumsTo100(milestones: readonly Milestone[] = SCHEDULE): boolean {
  return milestones.reduce((total, milestone) => total + milestone.pct, 0) === 100;
}

File: server/src/Backhaul.Domain/Money/Escrow.cs
using Backhaul.Domain.Trips;

namespace Backhaul.Domain.Money;

public enum MilestoneKind
{
    Advance,
    InTransit,
    Delivered,
    Retention,
}

/// <summary>One step of the release schedule.</summary>
public sealed record Milestone(MilestoneKind Kind, int Pct, string Condition);

/// <summary>What the platform knows when a release is being decided.</summary>
public sealed record EscrowConditions(
    TripState State,
    long MovingForMs,
    bool PodSealed,
    DateTimeOffset? DeliveredAt,
    bool ExceptionRaised);

/// <summary>A milestone, what it is worth on this trip, and whether it has been met.</summary>
public sealed record Release(Milestone Milestone, Kobo Amount, bool Met);

/// <summary>
/// When the money moves.
/// </summary>
/// <remarks>
/// Mirrors <c>packages/domain/src/escrow.ts</c>. Every condition reads
/// something the tracker, the trip machine or the proof engine produced —
/// never an opinion. A milestone that releases on a phone call is a milestone
/// that releases on an argument.
/// </remarks>
public static class Escrow
{
    /// <summary>How long the retention is held.</summary>
    public static readonly int RetentionDays = 7;

    /// <summary>Six hours of arriving positions before the second milestone releases.</summary>
    public static readonly long InTransitMs = 6 * 60 * 60_000L;

    public static readonly IReadOnlyList<Milestone> Schedule =
    [
        new(MilestoneKind.Advance, 30, "The truck reached the depot and loading started."),
        new(MilestoneKind.InTransit, 20, "The trip has been moving with positions arriving for six hours."),
        new(MilestoneKind.Delivered, 40, "Proof of delivery captured: photographs, a signature and a name."),
        new(MilestoneKind.Retention, 10, "Seven days after delivery with no exception raised."),
    ];

    public static bool IsMet(MilestoneKind kind, EscrowConditions conditions, DateTimeOffset now)
    {
        var started = conditions.State != TripState.Open && conditions.State != TripState.Assigned;

        switch (kind)
        {
            case MilestoneKind.Advance:
                return started;

            case MilestoneKind.InTransit:
                return started && conditions.MovingForMs >= InTransitMs;

            // Not the `delivered` state — the *proof*. A state is a claim
            // somebody made; the proof is photographs, a signature and a
            // position.
            case MilestoneKind.Delivered:
                return conditions.PodSealed;

            case MilestoneKind.Retention:
            {
                if (conditions.DeliveredAt is not { } delivered || !conditions.PodSealed) return false;
                // An open exception holds the retention. That is the entire
                // reason it exists, and releasing it on a timer regardless
                // would make it theatre.
                if (conditions.ExceptionRaised) return false;
                var elapsed = (now - delivered).TotalMilliseconds;
                return elapsed >= RetentionDays * 86_400_000L;
            }

            default:
                return false;
        }
    }

    /// <summary>Every milestone, met or not, with what each is worth.</summary>
    /// <remarks>
    /// A schedule that showed only what has been released would answer "how
    /// much have I had" and never "when do I get the rest", which is the
    /// question a carrier is actually asking.
    /// </remarks>
    public static IReadOnlyList<Release> For(Kobo agreed, EscrowConditions conditions, DateTimeOffset now) =>
        Schedule
            .Select(milestone => new Release(
                milestone,
                agreed.Percent(milestone.Pct),
                IsMet(milestone.Kind, conditions, now)))
            .ToList();

    public static Kobo Released(IReadOnlyList<Release> releases) =>
        releases.Where(r => r.Met).Aggregate(Kobo.Zero, (total, r) => total + r.Amount);

    public static Kobo HeldBack(Kobo agreed, IReadOnlyList<Release> releases) =>
        agreed - Released(releases);

    /// <summary>The next thing that has to happen for money to move, or null.</summary>
    public static Release? NextRelease(IReadOnlyList<Release> releases) =>
        releases.FirstOrDefault(r => !r.Met);

    /// <summary>
    /// Whether the schedule adds up.
    /// </summary>
    /// <remarks>
    /// Asserted rather than assumed: a schedule that sums to 95 quietly keeps
    /// 5% of every trip, and nobody would notice for months.
    /// </remarks>
    public static bool SumsTo100(IReadOnlyList<Milestone>? milestones = null) =>
        (milestones ?? Schedule).Sum(m => m.Pct) == 100;
}
